routing/DELEGATED_CONTENT_ROUTING_HTTP.md

```diff
@@ -0,0 +1,184 @@
+- Multiaddrs are string-encoded according to the [human-readable multiaddr specification](https://github.com/multiformats/multiaddr#specification)
+- Peer IDs are string-encoded according [PeerID string representation specification](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#string-representation)
+- Multibase bytes are string-encoded according to [the Multibase spec](https://github.com/multiformats/multibase), and *should* use base64.
+- Timestamps are Unix millisecond epoch timestamps
+
+Until required for business logic, servers should treat these types as opaque strings, and should preserve unknown JSON fields.
+
+### Versioning
+
+This API uses a standard version prefix in the path, such as `/v1/...`. If a backwards-incompatible change must be made, then the version number should be increased.
+
+### Provider Records
+
+A provider record contains information about a content provider, including the transfer protocol and any protocol-specific information useful for fetching the content from the provider.
+
+The information required to write a record to a router (*"write" provider records*) may be different than the information contained when reading provider records (*"read" provider records*).
+
+For example, indexers may require a signature in `bitswap` write records for authentication of the peer contained in the record, but the read records may not include this authentication information.
+
+Both read and write provider records have a minimal required schema as follows:
+
+```json
+{
+    "Protocol": "<transfer_protocol_name>",
+    "Schema": "<transfer_protocol_schema>",
+    ...
+}
+```
+
+Where:
+
+- `Protocol` is the multicodec name of the transfer protocol
+- `Schema` denotes the schema to use for encoding/decoding the record
+  - This is separate from the `Protocol` to allow this HTTP API to evolve independently of the transfer protocol
+  - Implementations should switch on this when parsing records, not on `Protocol`
+- `...` denotes opaque JSON, which may contain information specific to the transfer protocol
+
+Specifications for some transfer protocols are provided in the "Transfer Protocols" section.
+
+## API
+
+### `GET /routing/v1/providers/{CID}`
+
+- Response codes
+  - `200`: the response body contains 0 or more records
+  - `404`: must be returned if no matching records are found
+  - `422`: request does not conform to schema or semantic constraints
+- Response Body
+  
+  ```json
+  {
+      "Providers": [
+          {
+              "Protocol": "<protocol_name>",
+              "Schema": "<schema>",
+              ...
+          }
+      ]
+  }
+  ```
+  
+- Default limit: 100 providers
+- Optional query parameters
+  - `transfer` only return providers who support the passed transfer protocols, expressed as a comma-separated list of transfer protocol names such as `transfer=bitswap,filecoin-graphsync-v1`
```

Comment:
Just want to flag that `transfer` parameter is a hardcoded list of pre-vetted strings, which is making future changes extremely difficult:
-  it will be impossible to query for providers that expose novel third-party protocols
- every new protocol needs to be manually added to the spec here, then to libraries, and then rolled out to indexers and clients

Feels like gatekeeping system which makes it impossible to experiment with novel transfer protocols.
 
Last year I suggested  to not invent own strings, avoid transport protocol codes from [`table.csv`](https://github.com/multiformats/multicodec/blob/master/table.csv), but ensure clients can talk the same version of bitswap by reusing strings from  [libp2p identify's protocols list](https://github.com/libp2p/specs/tree/master/identify#protocols):

```console
$ ipfs id | jq .Protocols
[
  "/ipfs/bitswap",
  "/ipfs/bitswap/1.0.0",
  "/ipfs/bitswap/1.1.0",
  "/ipfs/bitswap/1.2.0",
  "/libp2p/fetch/0.0.1",
  ...
]
```

Since `transfer` is not implemented yet, I suggest removing it, or adding escape hatch for third-party protocols not specified in this document.

We could update spec and say that every value that starts with `/` should be interpreted as libp2p protocol prefix:

`transfer=/ipfs/bitswap` would return all peers that speak any version of bitswap.

Reply:
IMO using the numbers from the global table feels pretty bad for experimentation as well. The existing IPNI implementation doesn't support them and they require some gatekeeping anyway. For the time being I've put up https://github.com/ipni/specs/pull/6.

It seems like we could reasonably integrate this with something like the `/` proposal. 

> transfer=/ipfs/bitswap would return all peers that speak any version of bitswap.

Unfortunately, this type of query isn't supported in IPNI today and if it did it'd probably be with some specific semantics (e.g. adding a field to the "BitswapMetadata" section that's a list of protocolIDs) or a similar thing for GraphSync, unless they're going to reserve a new number every time they modify the protocol).

If we started leveraging named-record in addition to (or instead of) numbers in the global table we could just query against those. Basically, this would mean that the query could be fulfilled using custom logic per-number, or just using the `named-record` wrapper.

> Since transfer is not implemented yet, I suggest removing it

Since the general policy is not to merge specs without implementations we should remove and then re-add later with the implementation, right?

Reply:
> experimentation
>  require some gatekeeping

IPNI imposes no restriction on the protocol ID in metadata; this can be any number and treated as arbitrary bytes by the indexers. 

> Unfortunately, this type of query isn't supported in IPNI today

Happy to capture an issue on this if this is needed?

> if it did it'd probably be with some specific semantics

It'd probably be varint prefix matching.

> If we started leveraging named-record in addition to (or instead of) numbers in the global table we could just query against those

i'm sorry i am strugging to see how using names instad of numbers would make a difference in this case. Surely we can do the same with numbers?

Reply:
I think the consensus is to remove this, but the idea was that these are treated by this spec as opaque strings used for filtering, and the servers have no requirement to enforce any particular values (unless they want/need to). But happy to have this conversation another day, since we never implemented this anyway.